integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt

```diff
@@ -540,12 +569,100 @@ class ListenableFutureTest : TestBase() {
         finish(3)
     }
 
+    @Test
+    fun testFutureScopeCancellationPropagatedAlong() = runTest {
+        expect(1)
+        val supervisorJob = SupervisorJob()
+        val future = future(context = supervisorJob) {
+            expect(2)
+            try {
+                delay(Long.MAX_VALUE)
+                expectUnreached()
+            } catch (e: CancellationException) {
+                expect(4)
+                throw e
+            }
+        }
+        yield()
+        expect(3)
+        supervisorJob.cancel(CancellationException("Parent cancelled", TestException()))
+        supervisorJob.join()
+        assertTrue(future.isDone)
+        assertTrue(future.isCancelled)
+        assertFailsWith<CancellationException> { future.get() }.run {
+            assertEquals("Parent cancelled", message)
+            assertTrue(cause is TestException)
+        }
+        finish(5)
+    }
+
+    @Test
+    fun testFutureChildException() = runTest {
+        val future = future(context = NonCancellable + Dispatchers.Unconfined) {
+            val foo = async { delay(Long.MAX_VALUE); 42 }
+            val bar = async<Int> { throw TestException() }
+            foo.await() + bar.await()
+        }
+        future.checkFutureException<TestException>()
+    }
+
+    @Test
+    fun testFutureIsDoneAfterChildrenCompleted() = runTest {
+        expect(1)
+        // Don't propagate exception to the test and use different dispatchers as we are going to block test thread.
+        val future = future(context = NonCancellable + Dispatchers.Default) {
+            val foo = async {
+                try {
+                    delay(Long.MAX_VALUE)
+                    42
+                } finally {
+                    withContext(NonCancellable) {
+                        delay(200)
+                    }
+                }
+            }
+            foo.invokeOnCompletion {
+                expect(3)
+            }
+            val bar = async<Int> { throw TestException() }
```

Comment:
To strengthen the `is` below, create a `TestException` `val` scoped to the `runTest {}` block, throw that object instance here, and use object equality to assert that the expected exception was thrown, rather than a type introspection.

It's not likely to be some other `TestException` in this case, but it's the stronger statement, and it's *really* asserting that the expected exception was thrown. It's important with `IllegalStateException` in more complex cases.